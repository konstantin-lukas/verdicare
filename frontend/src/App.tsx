import React, {ReactNode, createContext, useState, useContext, useEffect} from 'react';

export const languages: { lang: string, flag: string }[] = [
    { lang: "bg", flag: "bg" },   // Bulgarian
    { lang: "cs", flag: "cz" },   // Czech
    { lang: "da", flag: "dk" },   // Danish
    { lang: "de", flag: "de" },   // German
    { lang: "el", flag: "gr" },   // Greek
    { lang: "en-gb", flag: "gb" },   // English (British)
    { lang: "en-us", flag: "us" },   // English (American)
    { lang: "es", flag: "es" },   // Spanish
    { lang: "et", flag: "ee" },   // Estonian
    { lang: "fi", flag: "fi" },   // Finnish
    { lang: "fr", flag: "fr" },   // French
    { lang: "hu", flag: "hu" },   // Hungarian
    { lang: "id", flag: "id" },   // Indonesian
    { lang: "it", flag: "it" },   // Italian
    { lang: "ja", flag: "jp" },   // Japanese
    { lang: "ko", flag: "kr" },   // Korean
    { lang: "lt", flag: "lt" },   // Lithuanian
    { lang: "lv", flag: "lv" },   // Latvian
    { lang: "nb", flag: "no" },   // Norwegian (BokmÃ¥l)
    { lang: "nl", flag: "nl" },   // Dutch
    { lang: "pl", flag: "pl" },   // Polish
    { lang: "pt-br", flag: "br" },   // Portuguese (Brazilian)
    { lang: "pt-pt", flag: "pt" },   // Portuguese (all Portuguese varieties excluding Brazilian Portuguese)
    { lang: "ro", flag: "ro" },   // Romanian
    { lang: "ru", flag: "ru" },   // Russian
    { lang: "sk", flag: "sk" },   // Slovak
    { lang: "sl", flag: "si" },   // Slovenian
    { lang: "sv", flag: "se" },   // Swedish
    { lang: "tr", flag: "tr" },   // Turkish
    { lang: "uk", flag: "ua" },   // Ukrainian
    { lang: "zh", flag: "cn" }    // Chinese (simplified)
];

function getInitialLanguage() {
    const language = localStorage.getItem('language');
    if (language === null) {
        const browserLang = navigator?.language?.toLowerCase();
        if (browserLang) {
            const foundLanguage = languages.find(lang => {
                return browserLang === lang.lang;
            });
            if (typeof foundLanguage !== 'undefined')
                return foundLanguage;
        }
        return languages[6];
    }
    return JSON.parse(language);
}
export let LanguageContext = createContext<{ lang: string, flag: string }>(getInitialLanguage());
export let SetLanguageContext = createContext<(language: { lang: string, flag: string }) => void>(() => {});
export let QueryContext = createContext<string>('');
export let SetQueryContext = createContext<(arg: string) => void>(() => {});
export let IsTypingContext = createContext<boolean>(false);
export let SetIsTypingContext = createContext<(arg: boolean) => void>(() => {});


export default function App({children}: {children: ReactNode[] | ReactNode}) {

    const [query, setQuery] = useState(useContext(QueryContext));
    const [typing, setTyping] = useState(useContext(IsTypingContext));
    const [currentLang, setCurrentLang] = useState(useContext(LanguageContext));
    useEffect(() => {
        localStorage.setItem('language', JSON.stringify(currentLang));
    }, [currentLang]);
    return (
        <LanguageContext.Provider value={currentLang}>
            <SetLanguageContext.Provider value={setCurrentLang}>
                <QueryContext.Provider value={query}>
                    <SetQueryContext.Provider value={setQuery}>
                        <IsTypingContext.Provider value={typing}>
                            <SetIsTypingContext.Provider value={setTyping}>
                                {children}
                           </SetIsTypingContext.Provider>
                        </IsTypingContext.Provider>
                    </SetQueryContext.Provider>
                </QueryContext.Provider>
            </SetLanguageContext.Provider>
        </LanguageContext.Provider>
    );
}

