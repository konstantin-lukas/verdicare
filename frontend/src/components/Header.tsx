import React, {useState, useContext, FormEvent} from 'react';
import {
    LanguageContext,
    SetLanguageContext,
    languages,
    SetQueryContext,
    SetIsTypingContext
} from '../App';
import logo from '../media/verdicare.svg'
import { HiMiniMagnifyingGlass } from 'react-icons/hi2';
import './Header.scss';
import {Link, useNavigate} from "react-router-dom";

function Languages({closeMenuCallback}: {closeMenuCallback:() => void}) {
    const currentLang = useContext(LanguageContext);
    const setCurrentLang = useContext(SetLanguageContext);
    const listItems = languages.filter(lang => {
        return lang.lang !== currentLang.lang;
    }).map(lang => {
        return (
            <li key={lang.flag} data-testid={lang.flag} className="flagWrapper" onClick={() => {
                setCurrentLang(lang);
                closeMenuCallback();
            }}>
                <span className={`fi fi-${lang.flag} fis`}></span>
            </li>
        );
    });
    return <>{listItems}</>;
}

function SearchInput({language}: {language: string}) {
    let placeholder: string;
    const navigate = useNavigate();
    const setQuery = useContext(SetQueryContext);
    const isTyping = useContext(SetIsTypingContext);
    const [typingTimeout, setTypingTimeout] = useState<null | NodeJS.Timeout>(null);

    switch (language) {
        case 'bg':
            placeholder = 'Търсете растение';
            break;
        case 'cs':
            placeholder = 'Hledejte rostlinu';
            break;
        case 'da':
            placeholder = 'Søg efter en plante';
            break;
        case 'de':
            placeholder = 'Suche nach einer Pflanze';
            break;
        case 'el':
            placeholder = 'Ψάξτε για φυτό';
            break;
        case 'en-gb':
        case 'en-us':
            placeholder = 'Look for a plant';
            break;
        case 'es':
            placeholder = 'Busca una planta';
            break;
        case 'et':
            placeholder = 'Otsige taime';
            break;
        case 'fi':
            placeholder = 'Etsi kasvia';
            break;
        case 'fr':
            placeholder = 'Cherchez une plante';
            break;
        case 'hu':
            placeholder = 'Keresse növény';
            break;
        case 'id':
            placeholder = 'Cari tanaman';
            break;
        case 'it':
            placeholder = 'Cerca una pianta';
            break;
        case 'ja':
            placeholder = '植物を検索';
            break;
        case 'ko':
            placeholder = '식물 찾기';
            break;
        case 'lt':
            placeholder = 'Ieškokite augalo';
            break;
        case 'lv':
            placeholder = 'Meklējiet augu';
            break;
        case 'nb':
            placeholder = 'Søk etter en plante';
            break;
        case 'nl':
            placeholder = 'Zoek een plant';
            break;
        case 'pl':
            placeholder = 'Szukaj rośliny';
            break;
        case 'pt-br':
        case 'pt-pt':
            placeholder = 'Procure por uma planta';
            break;
        case 'ro':
            placeholder = 'Căutați o plantă';
            break;
        case 'ru':
            placeholder = 'Ищите растение';
            break;
        case 'sk':
            placeholder = 'Hľadajte rastlinu';
            break;
        case 'sl':
            placeholder = 'Poiščite rastlino';
            break;
        case 'sv':
            placeholder = 'Sök efter en växt';
            break;
        case 'tr':
            placeholder = 'Bir bitki arayın';
            break;
        case 'uk':
            placeholder = 'Шукайте рослину';
            break;
        case 'zh':
            placeholder = '寻找植物';
            break;
        default:
            placeholder = 'Look for a plant';
            break;
    }

    const handleInput = (e: FormEvent) => {
        setQuery((e.target as HTMLInputElement).value);
        if (typingTimeout) {
            clearTimeout(typingTimeout);
            setTypingTimeout(null);
        }
        isTyping(true);
        setTypingTimeout(setTimeout(() => {
            isTyping(false);
            navigate("/");
        }, 650));
    }

    return <input type="text" name="search" placeholder={placeholder} onInput={handleInput}/>;
}
export default function Header() {

    const language = useContext(LanguageContext);
    let [langSelectOpen, setLangSelectOpen] = useState<boolean>(false);
    function toggleMenu() {
        setLangSelectOpen(!langSelectOpen);
    }
    function closeMenu() {
        setLangSelectOpen(false);
    }

    document.addEventListener('mousedown', e => {
        const elem = document.querySelector('#languageMenu');
        if (elem && e.target instanceof Element && elem !== e.target && !elem.contains(e.target)) {
            e.stopPropagation();
            setLangSelectOpen(false);
        }
    });

    return (
        <header>
            <Link to="/">
                <img src={logo} alt="Verdicare"/>
            </Link>
            <div>
                <div className="searchWrapper">
                    <SearchInput language={language.lang}/>
                    <HiMiniMagnifyingGlass className="searchIcon"/>
                </div>
                <div id="languageMenu">
                    <span className="flagWrapper" onClick={toggleMenu}>
                        <span className={`selectedLanguage fi fi-${language.flag} fis`}></span>
                    </span>
                    <div className={`languageSelect${langSelectOpen ? ' showLangSelect' : ''}`}>
                        <ul>
                            <Languages closeMenuCallback={closeMenu}/>
                        </ul>
                    </div>
                </div>
            </div>
        </header>
    );
}