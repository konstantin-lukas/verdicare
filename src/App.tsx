import React, {ReactNode, createContext, useState, useContext, useEffect} from 'react';

export const languages: { lang: string, flag: string }[] = [
    { lang: "bg", flag: "bg" },   // Bulgarian
    { lang: "cs", flag: "cz" },   // Czech
    { lang: "da", flag: "dk" },   // Danish
    { lang: "de", flag: "de" },   // German
    { lang: "el", flag: "gr" },   // Greek
    { lang: "en-gb", flag: "gb" },   // English (British)
    { lang: "en-us", flag: "us" },   // English (American)
    { lang: "es", flag: "es" },   // Spanish
    { lang: "et", flag: "ee" },   // Estonian
    { lang: "fi", flag: "fi" },   // Finnish
    { lang: "fr", flag: "fr" },   // French
    { lang: "hu", flag: "hu" },   // Hungarian
    { lang: "id", flag: "id" },   // Indonesian
    { lang: "it", flag: "it" },   // Italian
    { lang: "ja", flag: "jp" },   // Japanese
    { lang: "ko", flag: "kr" },   // Korean
    { lang: "lt", flag: "lt" },   // Lithuanian
    { lang: "lv", flag: "lv" },   // Latvian
    { lang: "nb", flag: "no" },   // Norwegian (Bokm√•l)
    { lang: "nl", flag: "nl" },   // Dutch
    { lang: "pl", flag: "pl" },   // Polish
    { lang: "pt-br", flag: "br" },   // Portuguese (Brazilian)
    { lang: "pt-pt", flag: "pt" },   // Portuguese (all Portuguese varieties excluding Brazilian Portuguese)
    { lang: "ro", flag: "ro" },   // Romanian
    { lang: "ru", flag: "ru" },   // Russian
    { lang: "sk", flag: "sk" },   // Slovak
    { lang: "sl", flag: "si" },   // Slovenian
    { lang: "sv", flag: "se" },   // Swedish
    { lang: "tr", flag: "tr" },   // Turkish
    { lang: "uk", flag: "ua" },   // Ukrainian
    { lang: "zh", flag: "cn" }    // Chinese (simplified)
];

function getInitialLanguage() {
    const language = localStorage.getItem('language');
    if (language === null)
        return languages[6];
    return JSON.parse(language);
}
export let LanguageContext = createContext<{ lang: string, flag: string }>(getInitialLanguage());
export let SetLanguageContext = createContext<(language: { lang: string, flag: string }) => void>(() => {});


export default function App({children}: {children: ReactNode[] | ReactNode}) {

    let [currentLang, setCurrentLang] = useState(useContext(LanguageContext));
    useEffect(() => {
        localStorage.setItem('language', JSON.stringify(currentLang));
    }, [currentLang]);
    return (
        <LanguageContext.Provider value={currentLang}>
            <SetLanguageContext.Provider value={setCurrentLang}>
                {children}
            </SetLanguageContext.Provider>
        </LanguageContext.Provider>
    );
}

