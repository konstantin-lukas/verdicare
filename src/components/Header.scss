@import '../main';
@import "../../node_modules/flag-icons/css/flag-icons.min.css";

header {
  height: 200px;
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: 0 15vw;

  a {
    display: inline-block;
    height: 33%;

    img {
      height: 100%;
    }

  }

  > div {
    display: flex;
  }

  .searchWrapper {
    border-radius: 2em;
    @include mainBorder;
    padding: 0.5em 1.5em;
    position: relative;
    margin-right: 1.5em;
  }

  input[name=search] {
    font-size: 1rem;
    border: none;
    width: 20vw;
  }

  input[name=search]:not(:placeholder-shown) + .searchIcon {
    opacity: 0;
    right: 0;
  }

  .searchIcon {
    @include bezier(all);
    position: absolute;
    height: 1.4rem;
    width: 1.4rem;
    right: 1em;
    top: 1em;
    transform: translateY(-42%);
    color: $borderColor;
  }

  #languageMenu {
    position: relative;

    .flagWrapper {
      border-radius: 50%;
      display: block;
      width: calc(2.2rem + 2px);
      height: calc(2.2rem + 2px);
      position: relative;
      overflow: hidden;
      cursor: pointer;
      box-shadow: rgba(0,0,0,0.3) 0 0 20px;
    }

    .fi.fis {
      width: 100%;
      height: 100%;
      position: absolute;
    }

    .languageSelect {
      position: absolute;
      opacity: 0;
      visibility: hidden;
      background: white;
      width: calc(20vw + 4.6rem + 2px);
      @include mainBorder;
      right: 0;
      top: 4.4rem;
      border-radius: 1.2rem;
      padding: 1rem;
      ul {
        display: inline-grid;
        width: 100%;
        justify-items: center;
        grid-template-columns: repeat(6, 1fr);
        @media screen and (max-width: 1600px) {
          grid-template-columns: repeat(5, 1fr);
        }
        @media screen and (max-width: 1200px) {
          grid-template-columns: repeat(4, 1fr);
        }
        li {
          width: calc(2.2rem + 2px);
          height: calc(2.2rem + 2px);
          margin: .75rem;
        }
      }
    }




    .languageSelect.showLangSelect {
      @include bezier(all);
      opacity: 1;
      visibility: visible;
    }

  }

}