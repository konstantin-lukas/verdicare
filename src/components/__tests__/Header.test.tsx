import React from 'react';
import {fireEvent, render, screen} from '@testing-library/react';
import renderer from 'react-test-renderer'
import Header from '../Header';
import App from '../../App';
describe('Header', () => {

   it('renders the header component', () => {
       render(<Header/>);
       expect(screen.getByPlaceholderText('Look for a plant')).toBeInTheDocument();
   });
   it('can change the site language', () => {
       render(
           <App>
               <Header/>
           </App>
       );
       fireEvent.click(screen.getByTestId('de'));
       expect(screen.queryByPlaceholderText('Look for a plant')).not.toBeInTheDocument();
       expect(screen.getByPlaceholderText('Suche nach einer Pflanze')).toBeInTheDocument();
   });
   it('renders correctly', () => {
       const tree = renderer
           .create(
               <App>
                   <Header/>
               </App>
           ).toJSON();
       expect(tree).toMatchSnapshot();
   });
});