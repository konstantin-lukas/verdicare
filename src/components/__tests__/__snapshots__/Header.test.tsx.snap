// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders correctly 1`] = `
<header>
  <a
    href="/"
  >
    <img
      alt="Verdicare"
      src="verdicare.svg"
    />
  </a>
  <div>
    <div
      className="searchWrapper"
    >
      <input
        name="search"
        placeholder="Look for a plant"
        type="text"
      />
      <svg
        aria-hidden="true"
        className="searchIcon"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 20 20"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clipRule="evenodd"
          d="M9 3.5a5.5 5.5 0 100 11 5.5 5.5 0 000-11zM2 9a7 7 0 1112.452 4.391l3.328 3.329a.75.75 0 11-1.06 1.06l-3.329-3.328A7 7 0 012 9z"
          fillRule="evenodd"
        />
      </svg>
    </div>
    <div
      id="languageMenu"
    >
      <span
        className="flagWrapper"
        onClick={[Function]}
      >
        <span
          className="selectedLanguage fi fi-us fis"
        />
      </span>
      <div
        className="languageSelect"
      >
        <ul>
          <li
            className="flagWrapper"
            data-testid="bg"
            onClick={[Function]}
          >
            <span
              className="fi fi-bg fis"
            />
          </li>
          <li
            className="flagWrapper"
            data-testid="cz"
            onClick={[Function]}
          >
            <span
              className="fi fi-cz fis"
            />
          </li>
          <li
            className="flagWrapper"
            data-testid="dk"
            onClick={[Function]}
          >
            <span
              className="fi fi-dk fis"
            />
          </li>
          <li
            className="flagWrapper"
            data-testid="de"
            onClick={[Function]}
          >
            <span
              className="fi fi-de fis"
            />
          </li>
          <li
            className="flagWrapper"
            data-testid="gr"
            onClick={[Function]}
          >
            <span
              className="fi fi-gr fis"
            />
          </li>
          <li
            className="flagWrapper"
            data-testid="gb"
            onClick={[Function]}
          >
            <span
              className="fi fi-gb fis"
            />
          </li>
          <li
            className="flagWrapper"
            data-testid="es"
            onClick={[Function]}
          >
            <span
              className="fi fi-es fis"
            />
          </li>
          <li
            className="flagWrapper"
            data-testid="ee"
            onClick={[Function]}
          >
            <span
              className="fi fi-ee fis"
            />
          </li>
          <li
            className="flagWrapper"
            data-testid="fi"
            onClick={[Function]}
          >
            <span
              className="fi fi-fi fis"
            />
          </li>
          <li
            className="flagWrapper"
            data-testid="fr"
            onClick={[Function]}
          >
            <span
              className="fi fi-fr fis"
            />
          </li>
          <li
            className="flagWrapper"
            data-testid="hu"
            onClick={[Function]}
          >
            <span
              className="fi fi-hu fis"
            />
          </li>
          <li
            className="flagWrapper"
            data-testid="id"
            onClick={[Function]}
          >
            <span
              className="fi fi-id fis"
            />
          </li>
          <li
            className="flagWrapper"
            data-testid="it"
            onClick={[Function]}
          >
            <span
              className="fi fi-it fis"
            />
          </li>
          <li
            className="flagWrapper"
            data-testid="jp"
            onClick={[Function]}
          >
            <span
              className="fi fi-jp fis"
            />
          </li>
          <li
            className="flagWrapper"
            data-testid="kr"
            onClick={[Function]}
          >
            <span
              className="fi fi-kr fis"
            />
          </li>
          <li
            className="flagWrapper"
            data-testid="lt"
            onClick={[Function]}
          >
            <span
              className="fi fi-lt fis"
            />
          </li>
          <li
            className="flagWrapper"
            data-testid="lv"
            onClick={[Function]}
          >
            <span
              className="fi fi-lv fis"
            />
          </li>
          <li
            className="flagWrapper"
            data-testid="no"
            onClick={[Function]}
          >
            <span
              className="fi fi-no fis"
            />
          </li>
          <li
            className="flagWrapper"
            data-testid="nl"
            onClick={[Function]}
          >
            <span
              className="fi fi-nl fis"
            />
          </li>
          <li
            className="flagWrapper"
            data-testid="pl"
            onClick={[Function]}
          >
            <span
              className="fi fi-pl fis"
            />
          </li>
          <li
            className="flagWrapper"
            data-testid="br"
            onClick={[Function]}
          >
            <span
              className="fi fi-br fis"
            />
          </li>
          <li
            className="flagWrapper"
            data-testid="pt"
            onClick={[Function]}
          >
            <span
              className="fi fi-pt fis"
            />
          </li>
          <li
            className="flagWrapper"
            data-testid="ro"
            onClick={[Function]}
          >
            <span
              className="fi fi-ro fis"
            />
          </li>
          <li
            className="flagWrapper"
            data-testid="ru"
            onClick={[Function]}
          >
            <span
              className="fi fi-ru fis"
            />
          </li>
          <li
            className="flagWrapper"
            data-testid="sk"
            onClick={[Function]}
          >
            <span
              className="fi fi-sk fis"
            />
          </li>
          <li
            className="flagWrapper"
            data-testid="si"
            onClick={[Function]}
          >
            <span
              className="fi fi-si fis"
            />
          </li>
          <li
            className="flagWrapper"
            data-testid="se"
            onClick={[Function]}
          >
            <span
              className="fi fi-se fis"
            />
          </li>
          <li
            className="flagWrapper"
            data-testid="tr"
            onClick={[Function]}
          >
            <span
              className="fi fi-tr fis"
            />
          </li>
          <li
            className="flagWrapper"
            data-testid="ua"
            onClick={[Function]}
          >
            <span
              className="fi fi-ua fis"
            />
          </li>
          <li
            className="flagWrapper"
            data-testid="cn"
            onClick={[Function]}
          >
            <span
              className="fi fi-cn fis"
            />
          </li>
        </ul>
      </div>
    </div>
  </div>
</header>
`;
